File: SpriteKit/SKParticleEmitter/SKParticleEmitter.m
//
//  SKParticleEmitter.m
//  ParticleEmitterDemo-SK
//
//  Created by Tom Bradley on 10/02/2014.
//  Copyright (c) 2014 71Squared Ltd. All rights reserved.
//

#import "SKParticleEmitter.h"

@interface SKParticleEmitter() {
    SKUniform *_u;
}
@end

@implementation SKParticleEmitter

- (void)loadTexture {
    _texture = [SKTexture textureWithCGImage:_cgImage];
}

- (void)setupArrays {
    [super setupArrays];
    
    _u = [SKUniform uniformWithName:@"u_opacityModifyRGB"];
    
    NSString *shaderFile = [[NSBundle mainBundle] pathForResource:@"skParticleFragmentShader.fsh" ofType:nil];
    NSString *shaderString = [NSString stringWithContentsOfFile:shaderFile encoding:NSASCIIStringEncoding error:nil];
    _shader = [[SKShader alloc] initWithSource:shaderString];
    
    [_u setFloatValue:1];
    [_shader addUniform:_u];
    
    _emitter = [SKNode new];
    _particleNodes = [[NSMutableArray alloc] initWithCapacity:maxParticles];
    
    for (int i=0; i<maxParticles; i++) {
        SKSpriteNode *particleNode = [SKSpriteNode spriteNodeWithTexture:_texture];
        [particleNode setSize:CGSizeMake(0, 0)];
        particleNode.shader = _shader;
        [_particleNodes addObject:particleNode];
        [particleNode setHidden:YES];
        particleNode.color = [SKColor colorWithRed:1.0 green:0.3 blue:0.5 alpha:0.5];
        [_emitter addChild:particleNode];
    }
}

- (void)updateWithDelta:(GLfloat)aDelta {
    [super updateWithDelta:aDelta];
    
    SKSpriteNode *particleNode;
    
    for (int i=0; i<particleCount; i++) {
        particleNode = _particleNodes[i];
        Particle p = particles[i];
        [particleNode setPosition:CGPointMake(p.position.x,
                                              p.position.y)];
        [particleNode setSize:CGSizeMake(p.particleSize, p.particleSize)];
        particleNode.color = [SKColor colorWithRed:p.color.r green:p.color.g blue:p.color.b alpha:p.color.a];
        particleNode.zRotation = GLKMathDegreesToRadians(p.rotation);
        particleNode.colorBlendFactor = 1.0;
        particleNode.blendMode = SKBlendModeAdd;
    }
}

- (BOOL)addParticle {
    BOOL success = [super addParticle];
    
    if (success) {
        SKSpriteNode *particleNode = [_particleNodes objectAtIndex:particleNodeIndex];
        [particleNode setHidden:NO];
        
        particleNodeIndex++;
        assert(particleNodeIndex <= maxParticles);
    }
    
    return success;
}

- (void)removeParticleAtIndex:(int)index {
    [super removeParticleAtIndex:index];
    
    particleNodeIndex--;
    assert(particleNodeIndex >= 0);
    
    SKSpriteNode *particleNode = [_particleNodes objectAtIndex:particleNodeIndex];
    [particleNode setHidden:YES];
}

@end

File: SpriteKit/SKParticleEmitter/SKParticleEmitter.swift
//
//  SKParticleEmitter.swift
//  ParticleEmitterDemo-SK
//
//  This is a Swift port of the SKParticleEmitter by 71Squared.
//
//  Created by Peter Easdown on 27/3/21.
//  Copyright © 2021 71Squared Ltd. All rights reserved.
//

import Foundation
import CoreGraphics
import SpriteKit

class SKParticleEmitter : SKNode, BaseParticleEmitterDelegate {
    
    /// The actual emitter engine.
    var emitter : BaseParticleEmitter?
    
    /// The texture to use for this emitter.
    var texture : SKTexture?

    /// The index of the next particle to be added to the node tree.
    var particleNodeIndex : NSInteger = 0

    /// The array of particles; only particleNodeIndex-1 of these will be in the tree, and those are updated every frame with
    /// properties from the underlying engine.
    var particleNodes = Array<SKSpriteNode>()
    
    /// Initialises the emitter (an SKNode) using the configuration specified file.
    /// - Parameter fileName: the configuration file to use; it's assumed that the extension is ".pex".
    /// - Throws: If something prevents the shader configuration being loaded.
    init(withConfigFile fileName: String) throws {
        super.init()

        // load the emitter engine.
        emitter = try BaseParticleEmitter.load(withFile: fileName, delegate: self)!
        
        // get the texture from the emitter.
        self.texture = emitter?.textureDetails!.texture()!
        
        // create all of the particle SKSpriteNodes, all using the same texture.
        for _ in 0 ..< emitter!.maxParticles.int {
            let particleNode = SKSpriteNode(texture: self.texture)
            particleNode.size = .zero
            particleNode.blendMode = emitter!.spriteKitBlendMode
            self.particleNodes.append(particleNode)
            particleNode.isHidden = true
            particleNode.color = .init(red: 1.0, green: 0.3, blue: 0.5, alpha: 0.5)
            self.addChild(particleNode)
        }
    }
    
    /// Not used by this class, but required by Swift.
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    /// Update the emitter (typically, once a frame).  Call this from the scene or view update() method.
    /// - Parameter aDelta: the amount of time to give to the particles.
    func update(withDelta aDelta: TimeInterval) {
        self.emitter?.update(withDelta: PEFloat(aDelta))
        
        var particleNode : SKSpriteNode

        // for every particle currently alive in the emitter, update the respective SKSpriteNode
        // with the parameters of the particle.
        for pi in 0 ..< emitter!.particleCount {
            particleNode = particleNodes[pi]
            
            let p = emitter!.particles[pi]
            particleNode.position = CGPoint(x: CGFloat(p.position.x), y: CGFloat(p.position.y))
            particleNode.size = .init(width: CGFloat(p.particleSize.float), height: CGFloat(p.particleSize.float))
            particleNode.color = p.color.asUIColor()
            particleNode.zRotation = CGFloat(GLKMathDegreesToRadians(p.rotation.float))
            particleNode.colorBlendFactor = 1.0
            
            // this is what 71Squared had.  I think with time, we can improve on this to map the GL blend modes to
            // SpriteKit blend modes.
//            particleNode.blendMode = self.emitter!.spriteKitBlendMode
        }
    }
    
    // MARK: - BaseParticleEmitterDelegate
    
    /// Adds a single particle node to the node tree.  This is called by the emitter engine.
    ///
    func addParticle() {
        let particleNode = self.particleNodes[self.particleNodeIndex]
        particleNode.isHidden = false
        
        particleNodeIndex += 1
        assert(particleNodeIndex <= emitter!.maxParticles.int)
    }
    
    /// Removes a particle from the node tree.
    func removeParticle() {
        particleNodeIndex -= 1
        assert(particleNodeIndex >= 0)

        let particleNode = self.particleNodes[self.particleNodeIndex]
        particleNode.isHidden = true
    }

}
